compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -602,7 +602,47 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         }
 
-        debug!("For closure={:?}, min_captures={:#?}", closure_def_id, root_var_min_capture_list);
+        debug!(
+            "For closure={:?}, min_captures before sorting={:?}",
+            closure_def_id, root_var_min_capture_list
+        );
+
+        // Now that we have the minimized list of captures, sort the captures by field id.
+        // This causes the closure to capture the upvars in the same order as the fields are
+        // declared which is also the drop order. Thus, in situations where we capture all the
+        // fields of some type, the obserable drop order will remain the same as it previously
+        // was even though we're dropping each capture individually.
+        // See https://github.com/rust-lang/project-rfc-2229/issues/42 and
+        // `src/test/ui/closures/2229_closure_analysis/preserve_field_drop_order.rs`.
+        for (_, captures) in &mut root_var_min_capture_list {
+            captures.sort_by(|capture1, capture2| {
```

Comment:
We *could* use `sort_by_key(|c| &c.place.projections)`, right? It wouldn't detect the buggy cases, but it'd be a heck of a lot shorter!

Reply:
`hir::place::Projection` doesn't impl `Ord` so that won't currently work. The MIR version of `Projection` does impl Ord so that doesn't seem entirely unreasonable to do. Would you rather I do that and use `sort_by_key` or leave the code as-is? 🙂

Reply:
I'm of two minds. Thinking about it more, I thought it seems a bit *fragile* to have the runtime semantics be derived by Ord in this way. At minimum, I would want a comment. =)